tests/testsuite/install.rs

```diff
@@ -2287,3 +2327,84 @@ fn sparse_install() {
 "#,
     );
 }
+
+#[cargo_test]
+fn self_referential() {
+    // Some packages build-dep on prior versions of themselves.
+    Package::new("foo", "0.0.1")
+        .file("src/lib.rs", "fn hello() {}")
+        .file("src/main.rs", "fn main() {}")
+        .file("build.rs", "fn main() {}")
+        .publish();
+    Package::new("foo", "0.0.2")
+        .file("src/lib.rs", "fn hello() {}")
+        .file("src/main.rs", "fn main() {}")
+        .file("build.rs", "fn main() {}")
+        .build_dep("foo", "0.0.1")
+        .publish();
+
+    cargo_process("install foo")
+        .with_stderr(
+            "\
+[UPDATING] `[..]` index
+[DOWNLOADING] crates ...
+[DOWNLOADED] foo v0.0.2 (registry [..])
+[INSTALLING] foo v0.0.2
+[DOWNLOADING] crates ...
+[DOWNLOADED] foo v0.0.1 (registry [..])
+[COMPILING] foo v0.0.1
+[COMPILING] foo v0.0.2
+[FINISHED] release [optimized] target(s) in [..]
+[INSTALLING] [CWD]/home/.cargo/bin/foo[EXE]
+[INSTALLED] package `foo v0.0.2` (executable `foo[EXE]`)
+[WARNING] be sure to add `[..]` to your PATH to be able to run the installed binaries
+",
+        )
+        .run();
+    assert_has_installed_exe(cargo_home(), "foo");
+}
+
+#[cargo_test]
+fn ambiguous_registry_vs_local_workspace_package() {
+    // Correctly install 'foo' from a workspace, even if that workspace
```

Comment:
Thanks for adding more tests! However, I don't fully understand where this test came from? Was it from a regression you perceived? A dependency of a binary being installed via `--path` shouldn't affect the behavior of `cargo install` in general. Could you explain it a bit why it is worth added?

BTW, this project setup doesn't look like a workspace.

Reply:
This test fails without this PR, just like the `self_referential` test. (`cargo build -p foo` could means registry `foo` or local `foo`.)

It stems from the comment I made on the associated issue #11999. I felt it might be worth adding because this is an example of non-recursive dependencies. It's just a "workspace has package named `foo` and also somewhere a dependency on `foo` from registry." This is not necessarily _recursive_, so is a more "likely" ("reasonable"?) situation to support. Or in short: I could see someone saying "we shouldn't support recursive dependencies like that," but this test points out it's not necessarily recursive, it could just be ordinary name clashing.

While the workspace is sort of the motivating reason why this could happen and be a totally reasonable thing that should definitely be supported, a "real" workspace isn't necessary to trigger the issue, so I just created a singular package of the same name.

I believe the particular example I saw once in the wild was a workspace that named its integration test packages after the library it was testing against. This isn't a major problem for `cargo install` (because you don't install an integration test?) but it was affecting another tool I built that liked to `cargo build -p pkgname` and equally hit these ambiguities.

Reply:
Make a lot of sense!  I was forgetting what the original issue was. Thank you for adding this extra test to prevent regression!